### PR TITLE
[BugFix] Priority scheduling and spec tokens preemption

### PR DESCRIPTION
**Pull request by andylolu2**
## Purpose

There is a bug on the edge case of request preemption with speculative decoding enabled. Scheduled speculative tokens were not cleared when the request is preempted. Here is the fix.

## Test Plan

I find such bugs difficult to cover well in unit-test like situations, so I opted for more "e2e"-like testing, in hopes that the test will become more robust to future changes.

The test runs for ~10s per test case, and it only uses CPU so it's not that expensive.

## Side note 1

I think this preemption logic needs serious refactoring. I think it'll be better to have `self.running` as a `RequestQueue` instead of plain list, and `schedule` will iterate over `self.running` in the order of high-to-low priority. This way we won't run into the edge case of preempting something that has already been scheduled.

I tried to draft this refactor but it requires `RequestQueue` to support two-sided ordering (need to iterate from high-to-low priority but preemption on lowest priority). This is doable with two heaps but I'm not sure of the performance penalty, so leaving it out of this PR for now. 

## Side note 2

I find the stability of vLLM to be noticeably worse than ~2-3 months ago. There are many "bug fix" PRs which is indication that a random snapshot of main is not very stable. I also noticed that many of these "bug fixes" don't come with corresponding tests, which might add to the overall flakiness of the project.

### CODE REVIEW

**Review by njhill — APPROVED** [review]
Thanks @andylolu2!

And you make some good points, I will respond to them in a bit :)

File: vllm/v1/core/sched/scheduler.py [hunk 1]
```diff
@@ -300,6 +300,14 @@ def schedule(self) -> SchedulerOutput:
...
                             req_to_new_blocks.pop(preempted_req.request_id)
                             num_scheduled_tokens.pop(preempted_req.request_id)
+                            scheduled_spec_decode_tokens.pop(
+                                preempted_req.request_id, None
+                            )
+                            scheduled_encoder_inputs.pop(preempted_req.request_id, None)
+                            # TODO: To be optimal we should reset the
+                            # encoder_compute_budget here too.
+                            # (it should be very rare to preempt a prefilling
+                            # request though)
```
[the remark this answers is not in the document]
Comment (andylolu2) [reply vllm/v1/core/sched/scheduler.py]: Actually this is a good suggestion, accepting it. I didn't know `get_num_encoder_tokens` was a method.
